<?xml version="1.0"?>

<launch>

    <!-- ARGS -->
    <arg name="rvizfile"    default = "$(find gtsam_backend)/../../rviz/mvsec_test.rviz" />
    <arg name="config_path" default = "$(find gtsam_backend)/../../config/mvsec_vins_config.yaml" />
    <arg name="vins_folder" default = "$(find feature_tracker)/../" />

    <!-- RVIZ -->
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(arg rvizfile)"/>

    <!-- FRONT END! -->
    <node name="feature_tracker" pkg="feature_tracker" type="feature_tracker" output="log">
      <param name="config_file" type="string" value="$(arg config_path)" />
      <param name="vins_folder" type="string" value="$(arg vins_folder)" />
    </node>

    <!-- MASTER NODE! -->
    <node pkg="gtsam_backend" type="vio" name="vio" output="screen" clear_params="true" required="true">
      <remap from="/vio/data_imu" to="/visensor/imu" />
      <remap from="/vio/data_uv"  to="/feature_tracker/feature" />

      <param name="fixedId"  type="string" value="world" />
      <rosparam param="gravity">[0.0, 0.0, 9.81007]</rosparam>
      <param name="imuWait"  type="int" value="0" />
      <param name="featWait" type="int" value="5" />
      
      <!-- Rotation and Translation from camera frame to IMU frame -->
      <rosparam param="R_C0toI">[0.9999717314190615,    -0.007438121416209933,  0.001100323844221122,
                                 0.00743200596269379,    0.9999574688631824,    0.005461295826837418,
                                -0.0011408988276470173, -0.0054529638303831614, 0.9999844816472552]</rosparam>
      <rosparam param="p_IinC0">[-0.03921656415229387,   0.00621263233002485,   0.0012210059575531885]</rosparam>

      <!-- Initialization -->
      <rosparam param="prior_qGtoI">[0.716147, 0.051158, 0.133778, 0.683096]</rosparam>       
      <rosparam param="prior_pIinG">[0.925493, -6.214668, 0.422872]</rosparam>       
      <rosparam param="prior_vIinG">[1.128364, -2.280640, 0.213326]</rosparam>       
      <rosparam param="prior_ba">[0.00489771688759235973,  0.00800897351104824969, 0.03020588299505782420]</rosparam>       
      <rosparam param="prior_bg">[-0.00041196751646696610, 0.00992948457018005999, 0.02188282212555122189]</rosparam>       

      <param name="sigma_camera"                 type="double"   value="0.306555403" />
      
      <!-- Noise Values for ETH Dataset Sensor -->
      <param name="accelerometer_noise_density"  type="double"   value="0.08" />    <!-- sigma_a -->
      <param name="gyroscope_noise_density"      type="double"   value="0.004" />   <!-- sigma_g -->
      <param name="accelerometer_random_walk"    type="double"   value="0.00004" /> <!-- sigma_wa -->
      <param name="gyroscope_random_walk"        type="double"   value="2.0e-6" />  <!-- sigma_wg -->

      <!-- Noise Values for Initialization -->
      <param name="sigma_prior_rotation"         type="double"   value="0.1" />
      <param name="sigma_prior_translation"      type="double"   value="0.3" />
      <param name="sigma_velocity"               type="double"   value="0.1" />
      <param name="sigma_bias"                   type="double"   value="0.15" />
      <param name="sigma_pose_rotation"          type="double"   value="0.1" />
      <param name="sigma_pose_translation"       type="double"   value="0.2" />
    </node>

</launch>
